expr/shaper.go

```diff
@@ -603,10 +626,44 @@ func castType(zctx *zson.Context, inType, specType zng.Type) (zng.Type, error) {
 			return zctx.LookupTypeArray(out), nil
 		}
 		return zctx.LookupTypeSet(out), nil
-	default:
-		// Non-castable type pair with at least one
-		// (non-record) container: output column is left
-		// unchanged.
-		return inType, nil
 	}
+	if bestUnionSelector(specType, inType) > -1 {
+		return specType, nil
+	}
+	// Non-castable type pair with at least one
+	// (non-record) container: output column is left
+	// unchanged.
+	return inType, nil
+}
+
+// bestUnionSelector tries to return the most specific selector for t in
+// maybeUnion.  It returns -1 if maybeUnion is not a union or contains no type
+// compatible with t.  (Types are compatible if they have the same underlying
+// type.)  If maybeUnion contains t, bestUnionSelector returns its selector.
+// Otherwise, if maybeUnion contains t's underlying type, bestUnionSelector
+// returns its selector.  Finally, bestUnionSelector returns the smallest
+// maybeUnion selector whose type is compatible with t.
+func bestUnionSelector(maybeUnion, t zng.Type) int {
```

Comment:
I think it would be clear if you called these variables:
```suggestion
func bestUnionSelector(inType, specType zng.Type) int {
```

Reply:
(I switched the order per my comment above).